config/crds/vsphereproviderconfig_v1alpha1_vsphereclusterproviderconfig.yaml

```diff
@@ -26,16 +26,17 @@ spec:
           type: string
         metadata:
           type: object
+        vsphereCredentialSecret:
+          type: string
         vspherePassword:
           type: string
         vsphereServer:
           type: string
         vsphereUser:
           type: string
       required:
-      - vsphereUser
```

Comment:
is there a way to express that: vsphereUser, vspherePassword, vsphereCredentialSecret can not empty at the same time?

Reply:
Not that I know of. I looked at using the `oneof` but seems like this file is auto-generated and I don't know a good way to express that.
If you have some ideas please feel free to let me know.

Reply:
after a little bit research, i don't aware of anything like this in openAPI schema level.

with that, could you handle the situation at `GetVsphereCredentials()`, return error if all three are empty from usr input. currently? currently it returns with username, password both = nil, then fail implicitly later. which is not idea!

Reply:
Anything that we do to perform validation is good but also futile.  Even if we error out, the current machine controller doesn't know how to "kick out" of an error condition.  It will keep trying to recreate the machine on any error.  Ideally, we would have something like a validating webhook to check the validity of the spec and decide whether to allow it or not.  That way, the request wouldn't even get to the machine controller.  We could also define a standard set of errors that all providers follow and have the machine controller understand those errors.

Reply:
I agree a proper webhook implementation is what might be a good solution. Also, we need to have a proper status field similar to say how pods have. Where we could tell the user that the machine/cluster deployment has errored out with the appropriate error details. And when in such state the controller should stop retrying.

Reply:
i will write an issue for this, looks like we just have to live with the current situation.